### PR TITLE
feat: use pipeline for `reth stage unwind`

### LINKED ISSUE #6965: Use pipeline for `reth stage unwind`
## Problem

Currently, `reth stage unwind` relies on the `DatabaseProvider` for deleting the data data from the database https://github.com/paradigmxyz/reth/blob/1214f2df012b82d7296755d3baa675085450c5c8/bin/reth/src/commands/stage/unwind.rs#L71-L78

It's not ideal for two reasons:
1. It doesn't work with static files anymore. Static files can only be unwound using the pipeline stages, but `take_block_and_execution_range` works purely on the database.
2. `take_block_and_execution_range` is needed only for the blockchain tree (and this CLI), and adding the static files unwind logic there doesn't make a lot of sense.

## Solution

Invoke pipeline unwind in `reth stage unwind`.

### PR DESCRIPTION
**Pull request by PanGan21**
Closes: #6965

### CODE REVIEW

File: bin/reth/src/commands/stage/unwind.rs [hunk 1]
```diff
@@ -59,25 +61,35 @@ impl Command {
...
+            StaticFileProducer::new(
+                provider_factory.clone(),
+                provider_factory.static_file_provider(),
+                PruneModes::default(),
+            ),
+        );
 
-        println!("Unwound {} blocks", blocks_and_execution.len());
+        pipeline.unwind(*range.end(), None)?;
+        pipeline.run().await.map_err(|err| eyre::eyre!("Transaction error on unwind: {err}"))?;
```
Comment (shekhirin) [inline bin/reth/src/commands/stage/unwind.rs:90]: do we need to call `run` here?

File: bin/reth/src/commands/stage/unwind.rs [hunk 2]
```diff
@@ -59,25 +61,35 @@ impl Command {
...
+            provider_factory.clone(),
+            StaticFileProducer::new(
+                provider_factory.clone(),
+                provider_factory.static_file_provider(),
+                PruneModes::default(),
+            ),
+        );
 
-        println!("Unwound {} blocks", blocks_and_execution.len());
+        pipeline.unwind(*range.end(), None)?;
```
Comment (shekhirin) [inline bin/reth/src/commands/stage/unwind.rs:89]: i think it should rather be the `*range.start()`

File: bin/reth/src/commands/stage/unwind.rs [hunk 3]
```diff
@@ -59,25 +61,35 @@ impl Command {
...
         let range = self.command.unwind_range(&db)?;
 
         if *range.start() == 0 {
             eyre::bail!("Cannot unwind genesis block")
         }
 
-        let factory = ProviderFactory::new(&db, self.chain.clone(), data_dir.static_files_path())?;
-        let provider = factory.provider_rw()?;
+        let provider_factory =
+            ProviderFactory::new(db.clone(), self.chain.clone(), data_dir.static_files_path())?;
```
Comment (shekhirin) [inline bin/reth/src/commands/stage/unwind.rs:76]:
```suggestion
            ProviderFactory::new(db, self.chain.clone(), data_dir.static_files_path())?;
```

File: bin/reth/src/commands/stage/unwind.rs [hunk 4]
```diff
@@ -59,25 +61,35 @@ impl Command {
...
 
-        let blocks_and_execution = provider
-            .take_block_and_execution_range(&self.chain, range)
-            .map_err(|err| eyre::eyre!("Transaction error on unwind: {err}"))?;
+        let pipeline_builder = Pipeline::builder();
```
Comment (shekhirin) [inline bin/reth/src/commands/stage/unwind.rs:78]: i think it builds an empty pipeline with no stages, but we need to add a set of default stages to replicate the actual pipeline unwind

**Review by shekhirin — CHANGES_REQUESTED** [review]

File: bin/reth/src/commands/stage/unwind.rs (hunk 3 above)
Comment (PanGan21) [reply bin/reth/src/commands/stage/unwind.rs]: fixed

File: bin/reth/src/commands/stage/unwind.rs (hunk 2 above)
Comment (PanGan21) [reply bin/reth/src/commands/stage/unwind.rs]: fixed

File: bin/reth/src/commands/stage/unwind.rs (hunk 4 above)
Comment (PanGan21) [reply bin/reth/src/commands/stage/unwind.rs:72]: Does this mean that we have to implement what is implemented in `take_block_and_execution_range` in pipeline stages?

File: bin/reth/src/commands/stage/unwind.rs (hunk 1 above)
Comment (PanGan21) [reply bin/reth/src/commands/stage/unwind.rs]: removed

File: bin/reth/src/commands/stage/unwind.rs (hunk 4 above)
Comment (Rjected) [reply bin/reth/src/commands/stage/unwind.rs:72]: we might just need to use `DefaultStages`, I would check out how reth builds a pipeline normally and try to replicate that here

Comment (PanGan21) [reply bin/reth/src/commands/stage/unwind.rs:72]: Quick question. It seems I need a `TaskExecutor` in order to create a header and a body downloader for the `DefaultStages`. Is it fine to initialize in this function an `AsyncCliRunner` or I can should I pass it from a higher level? Thanks in advance

Comment (shekhirin) [reply bin/reth/src/commands/stage/unwind.rs:72]: you can see the example of initiating `DefaultStages` in https://github.com/paradigmxyz/reth/blob/main/bin/reth/src/commands/import.rs

Comment (PanGan21) [reply bin/reth/src/commands/stage/unwind.rs:72]: Import pipeline is using a `FileClient` to build the header and body downloaders. 
What type of client should I use?

Comment (Rjected) [reply bin/reth/src/commands/stage/unwind.rs:72]: > Import pipeline is using a `FileClient` to build the header and body downloaders. What type of client should I use?

The `FetchClient` should be used, although this tells me that we should have some sort of pipeline unwind method that does not require creating a network

**Comment by joshieDo** [comment]
thanks for the start @PanGan21 ! thinking better about the issue, there's a few more requirements, so will be taking over

**Comment by joshieDo** [comment]
from an holesky datadir up to block 1M, with 0 headers in mdbx:
* unwind 100 blocks ✅ 
* --debug.tip 1M_HASH ✅

**Comment by PanGan21** [comment]
Thanks @joshieDo for taking it over!

File: bin/reth/src/commands/stage/unwind.rs [hunk 5]
```diff
@@ -55,28 +85,150 @@ impl Command {
...
+            ProviderFactory::new(db, self.chain.clone(), data_dir.static_files_path())?;
 
+        let range = self.command.unwind_range(provider_factory.clone())?;
         if *range.start() == 0 {
             eyre::bail!("Cannot unwind genesis block")
         }
 
-        let factory = ProviderFactory::new(&db, self.chain.clone(), data_dir.static_files_path())?;
-        let provider = factory.provider_rw()?;
+        // Only execute a pipeline unwind if the start of the range touches the existing static
```
Comment (shekhirin) [inline bin/reth/src/commands/stage/unwind.rs:100]:
```suggestion
        // Only execute a pipeline unwind if the start of the range overlaps the existing static
```

**Review by Rjected — APPROVED** [review]
lgtm, tested this out and it works. Noting that on large ranges this will likely OOM, but the fix for that can be scoped out later

**Review by shekhirin — APPROVED** [review]

**Pull request merged**